x/ccv/provider/keeper/validator_set_update.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// DeleteEpochValidator removes epoch validator with `providerAddr` address
+func (k Keeper) DeleteEpochValidator(
+	ctx sdk.Context,
+	chainID string,
+	providerConsAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.EpochKey(chainID, providerConsAddr.ToSdkConsAddr()))
+}
+
+// DeleteAllEpochValidators deletes all the stored epoch validators
+func (k Keeper) DeleteAllEpochValidators(
+	ctx sdk.Context,
+	chainID string) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.EpochBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+
+	var keysToDel [][]byte
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		keysToDel = append(keysToDel, iterator.Key())
+	}
+	for _, delKey := range keysToDel {
+		store.Delete(delKey)
+	}
+}
+
+// IsEpochValidator returns `true` if the validator with `providerAddr` is set in and `false` otherwise
+func (k Keeper) IsEpochValidator(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.EpochKey(chainID, providerAddr.ToSdkConsAddr())) != nil
+}
+
+// GetAllEpochValidators returns all the epoch validators on chain `chainID`
+func (k Keeper) GetAllEpochValidators(
+	ctx sdk.Context,
+	chainID string) (optedInValidators []types.EpochValidator) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.EpochBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		iterator.Value()
+		var optedInValidator types.EpochValidator
+		if err := optedInValidator.Unmarshal(iterator.Value()); err != nil {
+			panic(fmt.Errorf("failed to unmarshal EpochValidator: %w", err))
+		}
+		optedInValidators = append(optedInValidators, optedInValidator)
+	}
+
+	return optedInValidators
+}
+
+// ComputeNextEpochValidators returns the next validator set that is responsible for validating consumer chain `chainID`,
+// based on the bonded validators.
+func (k Keeper) ComputeNextEpochValidators(
+	ctx sdk.Context,
+	chainID string,
+	bondedValidators []stakingtypes.Validator,
+) []types.EpochValidator {
+	var nextValidators []types.EpochValidator
+	for _, val := range bondedValidators {
+		// get next voting power and the next consumer public key
+		nextPower := k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
+		consAddr, err := val.GetConsAddr()
+		if err != nil {
+			// this should never happen but is recoverable if we exclude this validator from the `nextValidators`
+			k.Logger(ctx).Error("could not get consensus address of validator",
+				"validator", val.GetOperator().String(),
+				"error", err)
+			continue
+		}
+		nextConsumerPublicKey, foundConsumerPublicKey := k.GetValidatorConsumerPubKey(ctx, chainID, types.NewProviderConsAddress(consAddr))
+		if !foundConsumerPublicKey {
```

Comment:
It is verbose but my understanding is that we cannot use [format verbs](https://pkg.go.dev/fmt) such as `%+v` or `%s` with `k.Logger(ctx)`. For instance if we had:
```
k.Logger(ctx).Info("could not retrieve public key for validator (%+v) on consumer chain (%s)", "validator", "chainID")
```
this would log the following:
```
2:22PM INF could not retrieve public key for validator (%+v) on consumer chain (%s) module=x/ibc-provider
```
While, with the current approach, what gets logged is the following:
```
2:22PM INF could not retrieve public key for validator on consumer chain because the validator did not assign a new consumer key chainID=consu module=x/ibc-provider validator=cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng
```

Reply:
By verbose I mean we should use debug instead of info for logging.

Reply:
The `Info` level might be better because this way we would be able to ask the validator team to check their logs and we would have an understanding of which validators have not assigned a consumer key. We could then ping those validators to assign a consumer key if needed. With `Debug` level, this might not be so easy (?)